src/audio/pipewire.c

```diff
@@ -0,0 +1,419 @@
+    state->playback_sample_rate = 0;
+    message(3, "basic soundsystem initialisation completed without errors");
+    //  start the threaded loop
+    pw_thread_loop_start(state->loop);
+    return (AudioID *)state;
+}
+// configure a pipewire stream for the given speech dispatcher configuration and then prepair loop for playback
+static int pipewire_begin(AudioID *id, AudioTrack track)
+{
+    module_state *state = (module_state *)id;
+    const struct spa_pod *params;
+    uint8_t buffer[1024];
+    struct spa_pod_builder b = SPA_POD_BUILDER_INIT(buffer, sizeof(buffer));
+    int format;
+    message(3, "entering per-module configuration procedure");
+    if (track.bits == 16)
+    {
+        state->stride = track.num_channels * sizeof(uint16_t);
+        switch (state->id.format)
+        {
+        case SPD_AUDIO_LE:
+            format = SPA_AUDIO_FORMAT_S16_LE;
+            message(3, "audio format is signed 16 bit, little endian");
+            break;
+        case SPD_AUDIO_BE:
+            format = SPA_AUDIO_FORMAT_S16_BE;
+            message(3, "audio format is signed 16 bit, big endian");
+            break;
+        default:
+            error("invalid audio format specifier");
+            return -1;
+        }
+    }
+    else if (track.bits == 8)
+    {
+        state->stride = track.num_channels * sizeof(uint8_t);
+        format = SPA_AUDIO_FORMAT_S8;
+        message(3, "audio format is signed 8bit");
+    }
+    else
+    {
+        error("invalid audio format specifier");
+        return -1;
+    }
+    message(4, "current stride is: %i", state->stride);
+    message(3, "the sample rate provided by the module is: %i", track.sample_rate);
+    // is this the first time this function is called? then our sentinel value must still be there
+    if (state->playback_sample_rate == 0)
+    {
+        state->playback_sample_rate = track.sample_rate;
+    }
+    // if we went here before and we got a new rate from spd, aka a new module with a different sample rate connected
+    else if (state->playback_sample_rate != 0 && state->playback_sample_rate != track.sample_rate)
+    {
+        // then we disconnect the stream, letting the rest of the function reconnect it with the new values, since format was computed before
+        message(4, "backend initialised with different sample rate, another module connecting?");
+        message(4, "disconnecting the old stream now");
+        pw_stream_disconnect(state->stream);
+        // but don't forget to update the value of sample rate we cache, for next time
+        state->playback_sample_rate = track.sample_rate;
+    }
+    // otherwise, we went here before, but we got an identical sample rate, why are we called by spd? this must be a spurious wakeup
+    else
+    {
+        message(3, "warning: spurious call by speech dispatcher, sample rate is the same as what this backend was initialised with");
```

Comment:
It's not spurious, it's only normal: begin is called for each utterance, and end is called when it's over (i.e. spd will possibly not be sending audio for some time). When that happens, do not connect again, you can just re-use the existing connected stream. See how other audio backends do it.

Reply:
alright, corrected to something which should be more coherent. To clarify, doing nothing and returning is the best thing to do still, even if my comment was wrong?

Reply:
Yes